// convex/dist/cjs/cli/lib/localDeployment/utils.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var utils_exports = {};
__export(utils_exports, {
  LOCAL_BACKEND_INSTANCE_SECRET: () => LOCAL_BACKEND_INSTANCE_SECRET,
  chooseLocalBackendPorts: () => chooseLocalBackendPorts,
  choosePorts: () => choosePorts,
  printLocalDeploymentWelcomeMessage: () => printLocalDeploymentWelcomeMessage
});
module.exports = __toCommonJS(utils_exports);
var import_log = require("../../../bundler/log.js");
var import_detect_port = require("detect-port");
var import_chalk = require("chalk");
async function choosePorts(ctx, {
  count,
  requestedPorts,
  suggestedPorts,
  startPort
}) {
  const ports = [];
  for (let i = 0; i < count; i++) {
    const requestedPort = requestedPorts?.[i];
    if (requestedPort !== null) {
      const port = await (0, import_detect_port.detect)(requestedPort);
      if (port !== requestedPort) {
        return ctx.crash({
          exitCode: 1,
          errorType: "fatal",
          printedMessage: `Requested port ${requestedPort} is not available`
        });
      }
      ports.push(port);
    } else {
      const suggestedPort = suggestedPorts?.[ports.length] ?? null;
      if (suggestedPort !== null) {
        const port2 = await (0, import_detect_port.detect)(suggestedPort);
        if (port2 === suggestedPort) {
          ports.push(suggestedPort);
          continue;
        }
      }
      const portToTry = ports.length > 0 ? ports[ports.length - 1] + 1 : startPort;
      const port = await (0, import_detect_port.detect)(portToTry);
      ports.push(port);
    }
  }
  return ports;
}
async function chooseLocalBackendPorts(ctx, options) {
  const { suggestedPorts, requestedPorts } = options ?? {};
  const [cloudPort, sitePort] = await choosePorts(ctx, {
    count: 2,
    startPort: 3210,
    requestedPorts: [
      requestedPorts?.cloud ?? null,
      requestedPorts?.site ?? null
    ],
    suggestedPorts: [
      suggestedPorts?.cloud ?? null,
      suggestedPorts?.site ?? null
    ]
  });
  return { cloudPort, sitePort };
}
function printLocalDeploymentWelcomeMessage() {
  (0, import_log.logMessage)(
    import_chalk.chalkStderr.cyan("You're trying out the beta local deployment feature!")
  );
  (0, import_log.logMessage)(
    import_chalk.chalkStderr.cyan(
      "To learn more, read the docs: https://docs.convex.dev/cli/local-deployments"
    )
  );
}
const LOCAL_BACKEND_INSTANCE_SECRET = "4361726e697461732c206c69746572616c6c79206d65616e696e6720226c6974";
